Split OCL Grammar and Add Support for Invariants and Derived Attributes

david-xander opened the pull request:
### Pull Request Description  
This pull request reorganizes the OCL Antlr grammar by splitting the main `OCL.g4` file into two separate files:  

- **`OCL.g4`**: Contains the core OCL 2.4 grammar, focusing on the foundational aspects of OCL expressions.  
- **`UMLRSDS.g4`**: Builds upon the core OCL grammar and integrates it with the UML-RSDS specification tool, supporting UML 2.5 class diagram notation and OCL 2.4.  

### New Feature: Invariants and Derived Attributes  
This update introduces the ability to define invariants and derived attributes using the OCL grammar. Examples include:  

#### Invariant  
```ocl  
context MyContext inv:  
    history->last().notEmpty()
```  

#### Derived Attribute  
```ocl  
context MyClass::derivedAttribute: Integer  
    derive: attribute1 + attribute2  
```  

### Motivation  
The grammar split improves modularity, allowing developers to work with the OCL core independently or alongside the UML-RSDS tool. Additionally, the new support for invariants and derived attributes makes the grammar more expressive and powerful for practical modeling scenarios.

kaby76 commented:
I can no longer tell what is the start rule and top-level grammar.

You added a grammar UMLRSDS.g4. It has EOF-terminated rules `specification` and `expressions` which you moved from the original OCL.g4 to this grammar. These start rules test input files bondApp.km3 and expressionExamples.txt respectively. This seems fine.

However, you added to OCL.g4 two new EOF-terminated rules multipleContextSpecifications and contextSpecification, presumably to test examples/OCL/example.txt. I don't know which entry point you are using to parse example.txt.

The desc.xml has to be updated to reflect exactly what grammar/entry point is used to test what input files exactly.

kaby76 commented:
@kevinlano Please review. Ty.

david-xander commented:
Thank you for pointing out the issue regarding the start rules and `desc.xml` configuration. I’ve made the following updates to address the feedback:  

**Start Rules and Grammar Entry Points**  
   - The `UMLRSDS.g4` file retains the `specification` and `expressions` rules to test `bondApp.km3` and `expressionExamples.txt` respectively.  
   - I moved the `expressionExamples.txt` to the OCL test folder: `examples/OCL/expressionExamples.txt`.
   - The `multipleContextSpecifications` rules in `OCL.g4` have been reviewed and are now explicitly linked to parse `examples/OCL/contextExample.txt`.  

**Updated `desc.xml`**  
   - The `desc.xml` file has been updated to specify which is the entry point and the exact example to test it.

**Pushed Fixes**  
   - The changes have been pushed to the same branch associated with this pull request.  

Please let me know if there are any additional concerns or adjustments needed. I appreciate your time and feedback!

kaby76 commented:
Looks like contextExamples from `multipleContextSpecifications` does not parse.
```
03/22-21:29:09 ~/issues/g4-4451/ocl/Generated-CSharp-contextspecs
$ ./bin/Debug/net8.0/Test.exe ../examples/OCL/contextExamples.txt
line 17:8 no viable alternative at input 'contextMyClass::derivedAttribute'
CSharp 0 ../examples/OCL/contextExamples.txt fail 0.1179059
Total Time: 0.176188
03/22-21:29:32 ~/issues/g4-4451/ocl/Generated-CSharp-contextspecs
```

david-xander commented:
Thank you for pointing out the issue with the derived attribute parsing in the grammar. After reviewing the feedback, I realized the root cause was a mismatch between the syntax and the grammar rule. I had partially done these changes a few days ago, but the OCL.g4 file was updated last week, and I didn’t fully validate the adjustments after those updates. I apologize for the oversight and any inconvenience caused.

To resolve the issue, I introduced a context-specific rule for `qualified_name`:

```antlr
qualified_name
    : ENUMERATION_LITERAL
    ;
```

This ensures semantic clarity and aligns with the intended use of derived attributes without duplicating lexer rules. I also updated the `singleDerivedAttribute` rule accordingly:

```antlr
singleDerivedAttribute
    : 'context' qualified_name ':' type ('init:' expression)? 'derive:' expression
    ;
```

I’ve tested the fix with the provided example (`context MyClass::derivedAttribute`) and additional cases to confirm it resolves the issue without introducing regressions. I updated the example with a more complete derived attribute case. Everything now parses correctly as expected.

Thank you for catching this and providing valuable feedback. I’ll take extra care to validate changes like this more thoroughly in the future. Please let me know if there’s anything else that needs to be addressed.

kevinlano commented:
What is the relation between this revised OCL.g4 and the OMG grammar for OCL (or the Eclipse OCL grammar)?

david-xander commented:
Thank you for your question, Dr. Lano.

This revised `OCL.g4` grammar is a subset of the OCL grammar defined in the OMG specification, focusing exclusively on the OCL core. It remains consistent with the OMG concrete syntax for OCL. The `UML-RSDS.g4` grammar reuses this core grammar for OCL specifications but extends it to include UML modeling and elements not covered by the OCL core.  

The current update introduces support for "context" definitions specifically for derived attributes. This is a step toward broader support for other OCL constructs like `body`, `preconditions`, and `postconditions`. While this subset is limited, it enables compliance with a range of OCL examples and provides a foundation for further expansion.  

I included a detailed explanation of the relationship between this grammar and the original OMG OCL 2.4 Specification in the `README.md`. Additionally I added two references more for the UML-RSDS tool. If you think additional clarification is needed, I’d be happy to refine it further.

Let me know if this helps clarify things or if you’d like me to elaborate on any specific part.

kaby76 commented:
It seems you are making an extension to the original OCL.g4. Why didn't you just make an override/extension without modification to the original OCL.g4? It's possible but requires a different import .g4 structure.

david-xander commented:
Thank you for your comment. Let me clarify the intent behind this update.

The `OCL.g4` grammar in this PR is not an extension but a refinement of the original OCL grammar. My focus was on improving its support for core OCL constructs as defined in the OMG OCL 2.4 Specification. During this process, I discovered that the original grammar lacked support for several standard OCL features (e.g., `self->sides->forAll(p: Side | ...)`). These missing features have now been added to ensure compliance with the specification.

Additionally, this update involved splitting the grammar to enable the `OCL.g4` file to handle OCL-only code. The UML-specific elements, which were previously part of the original `OCL.g4`, have been moved to the `UML-RSDS.g4` file. This separation allows the `OCL.g4` to focus solely on OCL while enabling reuse in UML-related contexts through the `UML-RSDS.g4`.

I hope this clears up any confusion. Please let me know if further clarification is needed or if you’d like to discuss the import structure in more detail.

kevinlano commented:
Thanks,
that sounds fine.

Kevin
________________________________
From: David ***@***.***>
Sent: 23 March 2025 14:58
To: antlr/grammars-v4 ***@***.***>
Cc: Kevin Lano ***@***.***>; Mention ***@***.***>
Subject: Re: [antlr/grammars-v4] Split OCL Grammar and Add Support for Invariants and Derived Attributes (PR #4451)


Thank you for your question, Dr. Lano.

This revised OCL.g4 grammar is a subset of the OCL grammar defined in the OMG specification, focusing exclusively on the OCL core. It remains consistent with the OMG concrete syntax for OCL. The UML-RSDS.g4 grammar reuses this core grammar for OCL specifications but extends it to include UML modeling and elements not covered by the OCL core.

The current update introduces support for "context" definitions specifically for derived attributes. This is a step toward broader support for other OCL constructs like body, preconditions, and postconditions. While this subset is limited, it enables compliance with a range of OCL examples and provides a foundation for further expansion.

I included a detailed explanation of the relationship between this grammar and the original OMG OCL 2.4 Specification in the README.md. Additionally I added two references more for the UML-RSDS tool. If you think additional clarification is needed, I’d be happy to refine it further.

Let me know if this helps clarify things or if you’d like me to elaborate on any specific part.

—
Reply to this email directly, view it on GitHub<https://github.com/antlr/grammars-v4/pull/4451#issuecomment-2746258573>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABU4GKBGANFAOBZ7V74QNFT2V3D2ZAVCNFSM6AAAAABZR6WL3KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBWGI2TQNJXGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[david-xander]david-xander left a comment (antlr/grammars-v4#4451)<https://github.com/antlr/grammars-v4/pull/4451#issuecomment-2746258573>

Thank you for your question, Dr. Lano.

This revised OCL.g4 grammar is a subset of the OCL grammar defined in the OMG specification, focusing exclusively on the OCL core. It remains consistent with the OMG concrete syntax for OCL. The UML-RSDS.g4 grammar reuses this core grammar for OCL specifications but extends it to include UML modeling and elements not covered by the OCL core.

The current update introduces support for "context" definitions specifically for derived attributes. This is a step toward broader support for other OCL constructs like body, preconditions, and postconditions. While this subset is limited, it enables compliance with a range of OCL examples and provides a foundation for further expansion.

I included a detailed explanation of the relationship between this grammar and the original OMG OCL 2.4 Specification in the README.md. Additionally I added two references more for the UML-RSDS tool. If you think additional clarification is needed, I’d be happy to refine it further.

Let me know if this helps clarify things or if you’d like me to elaborate on any specific part.

—
Reply to this email directly, view it on GitHub<https://github.com/antlr/grammars-v4/pull/4451#issuecomment-2746258573>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABU4GKBGANFAOBZ7V74QNFT2V3D2ZAVCNFSM6AAAAABZR6WL3KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBWGI2TQNJXGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

david-xander commented:
Hi @kaby76, 

Up to this point, Dr. Lano reviewed and agreed the changes.
Is there something else I should do to complete the PR?

kaby76 commented:
@david-xander All fine with me. @teverett A build please. Ty.

david-xander commented:
It seems that the test case does not support paths with `/`, as in `/examples/OCL/contextExamples`. Should I fix this in the desc.xml? I used exactly the same reference that @kaby76 committed. Maybe the unit test chaged and now this only works without paths. Should I address it?

kaby76 commented:
There are several problems with the test generator in generating the correct test scripts because it makes incorrect assumptions. This is because I tried to make the tester compatible with the old Maven tester. I will need to update the test generator to work, but these steps can be used to get testing to work for now.

Please change the desc.xml to this:
```
<?xml version="1.0" encoding="UTF-8" ?>
<desc xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../_scripts/desc.xsd">
   <targets>Java</targets>
   <test>
      <name>spec</name>
      <entry-point>specification</entry-point>
      <inputs>examples/specification/*.km3</inputs>
   </test>
   <test>
      <name>expr</name>
      <entry-point>expressions</entry-point>
      <inputs>examples/expressions/*.txt</inputs>
   </test>
   <test>
      <name>contextspecs</name>
      <entry-point>multipleContextSpecifications</entry-point>
      <inputs>examples/multipleContextSpecifications/*.txt</inputs>
   </test>
</desc>
```
Please note the new directory names. Not only is the directory named after the entry point per test, but I also use a globbing pattern to allow for more than one input test file. Then, please move each of the tests into the respective directories named after the entry point. I.e.,
* `examples/UML-RSDS/bondApp.km3` => `examples/specification/bondApp.km3`
* `examples/OCL/expressionExamples.txt` => `examples/expressions/expressionExamples.txt`
* `examples/OCL/contextExamples.txt` => `examples/multipleContextSpecifications/contextExamples.txt`

Not only does this make the tests work, but it is better organized to understand what is being used to test what entry point.

david-xander commented:
@kaby76, thank you for your feedback. I fixed the `desc.xml` file and the directory names according to your specifications.

kaby76 commented:
@teverett Please start a Github Action for this PR. It should be fine now.

teverett commented:
@david-xander  thanks!
